crates/protocol/src/messages/events.rs

```diff
@@ -0,0 +1,129 @@
+//! Message event primitives for OP stack interoperability.
+//!
+//! <https://specs.optimism.io/interop/messaging.html#messaging>
+//! <https://github.com/ethereum-optimism/optimism/blob/34d5f66ade24bd1f3ce4ce7c0a6cfc1a6540eca1/packages/contracts-bedrock/src/L2/CrossL2Inbox.sol>
+use alloc::{vec, vec::Vec};
+use alloy_primitives::{keccak256, Address, Bytes, Log, U256};
+use alloy_sol_types::{sol, SolType};
+
+sol! {
+    /// @notice The struct for a pointer to a message payload in a remote (or local) chain.
+    #[derive(Default, Debug, PartialEq, Eq)]
+    struct MessageIdentifierAbi {
+        address origin;
+        uint256 blockNumber;
+        uint256 logIndex;
+        uint256 timestamp;
+        uint256 chainId;
+    }
+
+    /// @notice Emitted when a cross chain message is being executed.
+    /// @param msgHash Hash of message payload being executed.
+    /// @param id Encoded Identifier of the message.
+    #[derive(Default, Debug, PartialEq, Eq)]
+    event ExecutingMessage(bytes32 indexed msgHash, MessageIdentifierAbi id);
+
+    /// @notice Executes a cross chain message on the destination chain.
+    /// @param _id      Identifier of the message.
+    /// @param _target  Target address to call.
+    /// @param _message Message payload to call target with.
+    function executeMessage(
+        MessageIdentifierAbi calldata _id,
+        address _target,
+        bytes calldata _message
+    ) external;
+}
+
+/// A [MessagePayload] is the raw payload of an initiating message.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MessagePayload(Bytes);
+
+impl From<Log> for MessagePayload {
+    fn from(log: Log) -> Self {
+        let mut data = vec![0u8; log.topics().len() * 32 + log.data.data.len()];
+        for (i, topic) in log.topics().iter().enumerate() {
+            data[i * 32..(i + 1) * 32].copy_from_slice(topic.as_ref());
+        }
+        data[(log.topics().len() * 32)..].copy_from_slice(log.data.data.as_ref());
+        data.into()
+    }
+}
+
+impl From<Vec<u8>> for MessagePayload {
+    fn from(data: Vec<u8>) -> Self {
+        Self(Bytes::from(data))
+    }
+}
+
+impl From<Bytes> for MessagePayload {
+    fn from(bytes: Bytes) -> Self {
+        Self(bytes)
+    }
+}
+
+impl From<MessagePayload> for Bytes {
+    fn from(payload: MessagePayload) -> Self {
+        payload.0
+    }
+}
+
+impl AsRef<[u8]> for MessagePayload {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
```

Comment:
prefer if you can remove as much of this boiler plate code as possible deriving it instead
https://docs.rs/derive_more/latest/derive_more/all.html